sequencer/src/block2.rs

```diff
@@ -58,13 +66,168 @@ impl BlockPayload {
         payload.extend(tx_bodies);
         Self { payload }
     }
+
+    // Return the range `r` such that the `index`th tx bytes are at `self.payload[r]`.
+    fn get_tx_range(&self, index: TxIndex) -> Option<Range<usize>> {
+        let tx_bodies_offset = self.tx_bodies_offset()?;
+
+        // tx_table[i] is end index for the ith tx,
+        // so the range for the ith tx is tx_table[i-1..i].
+        // But tx_table starts at index 1 in the payload
+        // because index 0 in the payload is the length of the tx_table.
+        // So the range for the ith tx in the payload is actually
+        // [index..index+1].
+        let start = if index == 0 {
+            0
+        } else {
+            self.get_value_usize(index)?
+        }
+        .checked_add(tx_bodies_offset)?;
+
+        let end = self
+            .get_value_usize(index.checked_add(1)?)?
+            .checked_add(tx_bodies_offset)?;
+
+        Some(start..end)
+    }
+
+    // Viewing `self.payload` bytes as a vec of words of type `TxIndex`,
+    // return the `index`th word.
+    fn get_value(&self, index: TxIndex) -> Option<TxIndex> {
```

Comment:
Not sure I understand correctly. It feels a bit weird because both inputs and outputs are of type `TxIndex` but depending on the value passed in for `index` (if the index is larger than the table) the function could construct a TxIndex from tx payload data?

This function seems to be designed for looking up indexes in the table in the payload. Can we make this safer by indicating it in the name and making it panic or return an error if it's used wrongly?

Reply:
> It feels a bit weird because both inputs and outputs are of type `TxIndex` ...

Agree it's confusing. In an ideal world we would use Rust's type system to differentiate between an index *into* the tx table and the actualy *entry* itself. (I have a crusade for [typed collection indices](https://github.com/axelarnetwork/tofn/blob/4f2a4de3c4a57f4168368b2afd02bc8daaea6822/README.md#typed-collection-indices) for which I'm attempting to recruit adherents.)

This is an internal function so I did not take the time to make it bulletproof for downstream users.

> ...but depending on the value passed in for `index` (if the index is larger than the table) the function could construct a TxIndex from tx payload data?

Correct.

> This function seems to be designed for looking up indexes in the table in the payload. Can we make this safer by indicating it in the name and making it panic or return an error if it's used wrongly?

Not a bad idea. I'll do that.

Reply:
Introduced a newtype `TxTableEntry`, renamed `get_value` to `get_tx_table_entry`, added a check so that it can be used only to read from the tx table and nothing else.

(At first this function returned a `Option<TxTableEntry>` as one might expect but it's only use always converts to `usize` so I later moved the `TxTableEntry` -> `usize` conversion inside `gest_tx_table_entry`.)